### PR TITLE
Create Spearman's rank correlation coefficient

### PR DESCRIPTION
**Pull request by cyrixninja**
### Describe your change:
Implemented Python program to find Spearman's rank correlation coefficient. Spearman's rank correlation coefficient is a nonparametric measure of rank correlation (statistical dependence between the rankings of two variables). It assesses how well the relationship between two variables can be described using a monotonic function.

Wikipedia References: 
https://en.wikipedia.org/wiki/Spearman%27s_rank_correlation_coefficient

* [x] Add an algorithm?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.

### CODE REVIEW

File: maths/spearman_rank_correlation_coefficient.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
+def rankdata(arr):
```
[the remark this answers is not in the document]
Comment (cyrixninja) [reply maths/spearman_rank_correlation_coefficient.py:1]: Done

File: maths/spearman_rank_correlation_coefficient.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    ranked = sorted((value, index) for index, value in enumerate(arr))
+    ranks = [0] * len(arr)
+
+    for pos, (_, index) in enumerate(ranked):
+        ranks[index] = pos + 1
+
+    return ranks
+
+
+def spearman_rank_correlation(x, y):
```
[the remark this answers is not in the document]
Comment (cyrixninja) [reply maths/spearman_rank_correlation_coefficient.py:25]: Done

File: maths/spearman_rank_correlation_coefficient.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    ranked_data = sorted((value, index) for index, value in enumerate(data))
+    ranks = [0] * len(data)
+
+    for position, (_, index) in enumerate(ranked_data):
+        ranks[index] = position + 1
+
+    return ranks
+
+
+def calculate_spearman_rank_correlation(x: list[float], y: list[float]) -> float:
```
[the remark this answers is not in the document]
Comment (cyrixninja) [reply maths/spearman_rank_correlation_coefficient.py]: Done

File: maths/spearman_rank_correlation_coefficient.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+if __name__ == "__main__":
+    import doctest
+
+    doctest.testmod()
+
+    # Example usage:
+    x1 = [1, 2, 3, 4, 5]
+    y1 = [2, 4, 6, 8, 10]
+    rho1 = calculate_spearman_rank_correlation(x1, y1)
```
Comment (cclauss) [inline maths/spearman_rank_correlation_coefficient.py:75]:
```suggestion
    print(f"{calculate_spearman_rank_correlation([1, 2, 3, 4, 5], [2, 4, 6, 8, 10]) = }")
```

File: maths/spearman_rank_correlation_coefficient.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    print(f"Spearman's rank correlation coefficient (Example 1): {rho1}")
+
+    x2 = [1, 2, 3, 4, 5]
+    y2 = [5, 4, 3, 2, 1]
+    rho2 = calculate_spearman_rank_correlation(x2, y2)
```
Comment (cclauss) [inline maths/spearman_rank_correlation_coefficient.py:80]:
```suggestion
    print(f"{calculate_spearman_rank_correlation([1, 2, 3, 4, 5], [5, 4, 3, 2, 1]) = }")
```

File: maths/spearman_rank_correlation_coefficient.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    print(f"Spearman's rank correlation coefficient (Example 2): {rho2}")
+
+    x3 = [1, 2, 3, 4, 5]
+    y3 = [5, 1, 2, 9, 5]
+    rho3 = calculate_spearman_rank_correlation(x3, y3)
```
Comment (cclauss) [inline maths/spearman_rank_correlation_coefficient.py:85]:
```suggestion
    print(f"{calculate_spearman_rank_correlation([1, 2, 3, 4, 5], [5, 1, 2, 9, 5]) = }")
```

File: maths/spearman_rank_correlation_coefficient.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    print(f"Spearman's rank correlation coefficient (Example 3): {rho3}")
```
Comment (cclauss) [inline maths/spearman_rank_correlation_coefficient.py:86]:
```suggestion
```

File: maths/spearman_rank_correlation_coefficient.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    # Example usage:
+    x1 = [1, 2, 3, 4, 5]
+    y1 = [2, 4, 6, 8, 10]
+    rho1 = calculate_spearman_rank_correlation(x1, y1)
+    print(f"Spearman's rank correlation coefficient (Example 1): {rho1}")
+
+    x2 = [1, 2, 3, 4, 5]
+    y2 = [5, 4, 3, 2, 1]
+    rho2 = calculate_spearman_rank_correlation(x2, y2)
+    print(f"Spearman's rank correlation coefficient (Example 2): {rho2}")
```
Comment (cclauss) [inline maths/spearman_rank_correlation_coefficient.py:81]:
```suggestion
```

File: maths/spearman_rank_correlation_coefficient.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+if __name__ == "__main__":
+    import doctest
+
+    doctest.testmod()
+
+    # Example usage:
+    x1 = [1, 2, 3, 4, 5]
+    y1 = [2, 4, 6, 8, 10]
+    rho1 = calculate_spearman_rank_correlation(x1, y1)
+    print(f"Spearman's rank correlation coefficient (Example 1): {rho1}")
```
Comment (cclauss) [inline maths/spearman_rank_correlation_coefficient.py:76]:
```suggestion
```

**Review by cclauss — CHANGES_REQUESTED** [review]

File: maths/spearman_rank_correlation_coefficient.py [hunk 10]
```diff
@@ -0,0 +1,82 @@
+from collections.abc import Sequence
+
+
+def assign_ranks(data: list[float]) -> list[int]:
```
Comment (cclauss) [inline maths/spearman_rank_correlation_coefficient.py:4]:
```suggestion
def assign_ranks(data: Sequence[float]) -> list[int]:
```

**Review by cclauss — APPROVED** [review]

**Pull request merged**